Make sensitive jsc global namespace properties not enumerable by Object.getOwnPropertyNames.

MenloDorian opened the pull request:
#### 15477a1fd6b58f51f29ef8ee00e63b139ce49fb1
<pre>
Make sensitive jsc global namespace properties not enumerable by Object.getOwnPropertyNames.
<a href="https://bugs.webkit.org/show_bug.cgi?id=263506">https://bugs.webkit.org/show_bug.cgi?id=263506</a>
rdar://112815258

Reviewed by Alexey Shvayka and Justin Michaud.

Some functions in the jsc shell GlobalObject are only added as debugging aids.  They are meant
to be used carefully under controlled conditions for test development.  Though they are added
as DontEnum, Object.getOwnPropertyNames() still enumerates them.  We should filter out all
DontEnum properties of this GlobalObject so as not to trip up fuzzers that try to fuzz with
Object.getOwnPropertyNames.

Achieving this turned out to be somewhat challenging for the following reasons:
1. We want the jsc shell&apos;s GlobalObject to be allocated out of the same IsoHeap as JSGlobalObject.
2. To achieve (1), we cannot change the size of GlobalObject i.e. we cannot add a field.
3. In order to filter out the sensitive properties, we need to maintain a copy of this list of
   properties somewhere. We can&apos;t stash it in the GlobalObject.
4. The list consists of UniquedStringImpl pointers.  So, strictly speaking, we can stash it with
   the VM because UniquedStringImpls are unique to each VM.  However, this is an issue with the
   jsc shell&apos;s GlobalObject.  The jsc shell is by 1 client of the JSC framework.   We don&apos;t want
   the VM to have specific knowledge about the jsc shell.

To solve this, we&apos;re introducing a SideDataRepository that can be used to associate some side
data with some owner.  In this case, the owner here is the VM.  This SideDataRepository will also
come in handy for stashing side data for other slow or rarely used features that we don&apos;t want to
bloat commonly used data structures for.  For example, JIT comments could have been stashed as
side data.  JIT probe backing data can be stashed as side data.  For now, we&apos;re only using it to
store jsc shell&apos;s GlobalObject&apos;s property filter list.

We&apos;re still adding a VM::m_hasSideData bool so that we can optimize the VM destructor to avoid
checking for the need to clean up side data if non was ever added.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/jsc.cpp:
* Source/JavaScriptCore/runtime/SideDataRepository.cpp: Added.
(JSC::SideDataRepository::add):
(JSC::SideDataRepository::deleteAll):
(JSC::sideDataRepository):
* Source/JavaScriptCore/runtime/SideDataRepository.h: Added.
(JSC::SideDataRepository::ensure):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::~VM):
* Source/JavaScriptCore/runtime/VM.h:
* Source/JavaScriptCore/runtime/VMInlines.h:
(JSC::VM::ensureSideData):

Canonical link: <a href="https://commits.webkit.org/269639@main">https://commits.webkit.org/269639@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/594fe220a93f4c75f4d5e02a5b7ddff3cfacec05

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23141 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1226 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24260 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25059 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21397 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2687 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23679 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22252 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23381 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20065 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25908 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/642 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20951 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27118 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/20137 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/21081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21215 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24986 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22505 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/643 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18417 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/29879 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/574 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6177 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1052 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/29831 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2940 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/850 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6071 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/WTF/wtf/HashSet.h:
```diff
@@ -144,6 +144,11 @@ class HashSet final {
     template<typename OtherCollection>
     HashSet intersectionWith(const OtherCollection&) const;
 
+    // Returns a new set with the elements that are either in this set but not
```
I think this sentence makes sense without "either"?

justinmichaud reviewed (approved):
r=me

Commit:
Make sensitive jsc global namespace properties not enumerable by Object.getOwnPropertyNames.
https://bugs.webkit.org/show_bug.cgi?id=263506
rdar://112815258

Reviewed by Alexey Shvayka and Justin Michaud.

Some functions in the jsc shell GlobalObject are only added as debugging aids.  They are meant
to be used carefully under controlled conditions for test development.  Though they are added
as DontEnum, Object.getOwnPropertyNames() still enumerates them.  We should filter out all
DontEnum properties of this GlobalObject so as not to trip up fuzzers that try to fuzz with
Object.getOwnPropertyNames.

Achieving this turned out to be somewhat challenging for the following reasons:
1. We want the jsc shell's GlobalObject to be allocated out of the same IsoHeap as JSGlobalObject.
2. To achieve (1), we cannot change the size of GlobalObject i.e. we cannot add a field.
3. In order to filter out the sensitive properties, we need to maintain a copy of this list of
   properties somewhere. We can't stash it in the GlobalObject.
4. The list consists of UniquedStringImpl pointers.  So, strictly speaking, we can stash it with
   the VM because UniquedStringImpls are unique to each VM.  However, this is an issue with the
   jsc shell's GlobalObject.  The jsc shell is by 1 client of the JSC framework.   We don't want
   the VM to have specific knowledge about the jsc shell.

To solve this, we're introducing a SideDataRepository that can be used to associate some side
data with some owner.  In this case, the owner here is the VM.  This SideDataRepository will also
come in handy for stashing side data for other slow or rarely used features that we don't want to
bloat commonly used data structures for.  For example, JIT comments could have been stashed as
side data.  JIT probe backing data can be stashed as side data.  For now, we're only using it to
store jsc shell's GlobalObject's property filter list.

We're still adding a VM::m_hasSideData bool so that we can optimize the VM destructor to avoid
checking for the need to clean up side data if non was ever added.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/jsc.cpp:
* Source/JavaScriptCore/runtime/SideDataRepository.cpp: Added.
(JSC::SideDataRepository::add):
(JSC::SideDataRepository::deleteAll):
(JSC::sideDataRepository):
* Source/JavaScriptCore/runtime/SideDataRepository.h: Added.
(JSC::SideDataRepository::ensure):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::~VM):
* Source/JavaScriptCore/runtime/VM.h:
* Source/JavaScriptCore/runtime/VMInlines.h:
(JSC::VM::ensureSideData):

Canonical link: https://commits.webkit.org/269639@main
